source/module_hamilt_pw/hamilt_pwdft/kernels/rocm/stress_op.hip.cu

```diff
@@ -352,6 +352,87 @@ __global__ void cal_vq_deri(
 }
 
 
+
+template <typename FPTYPE>
+__global__ void cal_stress_drhoc_aux0(
+        const FPTYPE* r, const FPTYPE* rhoc, 
+        const FPTYPE *gx_arr, const FPTYPE *rab, FPTYPE *drhocg, 
+        const int mesh, const int igl0, const int ngg, const double omega
+){
+    const double FOUR_PI =  4.0 * 3.14159265358979323846;
+
+    int idx = threadIdx.x + blockIdx.x * blockDim.x;
+
+    FPTYPE aux_d[2];
+    FPTYPE rhocg1=0.0, f_0=0.0, f_2=0.0, f_1=0.0;
+
+    if (idx >= ngg) {return;}
+    
+    for( int ir = 0;ir< mesh; ir++)
+    {
+        aux_d [ir%2]  = r [ir] * rhoc [ir] * (r [ir] * cos (gx_arr[idx] * r [ir] ) / gx_arr[idx] - sin (gx_arr[idx] * r [ir] ) / pow(gx_arr[idx],2));
+
+        if(ir==0){
+            f_0 = aux_d[ir%2]*rab[ir];
+        } else if(ir==mesh-2){
+            f_2 = aux_d[ir%2]*rab[ir];
+        } else if(ir==mesh-1) {
+            f_1 = aux_d[ir%2]*rab[ir];
+        } else if(ir%2==0){
+            const double f1 = aux_d[1]*rab[ir-1];
+            rhocg1 += f1 + f1 + aux_d[0]*rab[ir];
+        }
+
+    }//ir
+    rhocg1 += f_2+f_2;
+    rhocg1 += rhocg1;
+    rhocg1 += f_0 + f_1;
+    rhocg1/=3.0;
+
+    drhocg [idx] = FOUR_PI / omega * rhocg1;
+}
+
+template <typename FPTYPE>
+__global__ void cal_stress_drhoc_aux1(
+        const FPTYPE* r, const FPTYPE* rhoc, 
+        const FPTYPE *gx_arr, const FPTYPE *rab, FPTYPE *drhocg, 
+        const int mesh, const int igl0, const int ngg, const double omega
+){
+    const double FOUR_PI =  4.0 * 3.14159265358979323846;
+
+    int idx = threadIdx.x + blockIdx.x * blockDim.x;
+
+    FPTYPE aux_d[2];
+    FPTYPE rhocg1=0.0, f_0=0.0, f_2=0.0, f_1=0.0;
+
+    if (idx >= ngg) {return;}
+    
+    for( int ir = 0;ir< mesh; ir++)
+    {
+        aux_d [ir%2] = ir!=0 ? sin(gx_arr[idx] * r[ir]) / (gx_arr[idx] * r[ir]) : 1.0;
+        aux_d [ir%2] = r[ir] * r[ir] * rhoc [ir] * aux_d [ir%2];
+
+        if(ir==0){
+            f_0 = aux_d[ir%2]*rab[ir];
+        } else if(ir==mesh-2){
+            f_2 = aux_d[ir%2]*rab[ir];
+        } else if(ir==mesh-1) {
+            f_1 = aux_d[ir%2]*rab[ir];
+        } else if(ir%2==0){
+            const double f1 = aux_d[1]*rab[ir-1];
+            rhocg1 += f1 + f1 + aux_d[0]*rab[ir];
+        }
```

Comment:
If the ir cycle is combined, then the three boundary conditions regarding the `mesh-1`,`mesh-2` and the `0` point need to be unpacked outside the cycle. This will result in a lot of duplicate code and affect the readability of the code.

And due to compiler optimization, this writing does not cause warp differences at present, and has little impact on performance (after testing, the kernel time before and after changing in this way is almost the same (about `0.08+-0.02s`)).

So I tend not to make these change.

Reply:
You can use a lambda expression to avoid duplicate code and make the program logic more clear, and the modified code snippet of function `cal_stress_drhoc_aux0` is as follows: 
```
    FPTYPE rhocg1=0.0, f_0=0.0, f_2=0.0, f_1=0.0;
    FPTYPE gx = gx_arr[idx]
    auto aux = [](FPTYPE r, FPTYPE rhoc, FPTYPE gx_arr, FPTYPE rab) -> FPTYPE {
        return r * rhoc * (r * cos (gx_arr * r) / gx_arr - sin (gx_arr * r) / (gx_arr * gx_arr)) * rab;
    };

    if (idx >= ngg) {return;}
    
    f_0 = aux(r[0], rhoc[0], gx, rab[0]);
    for( int ir = 1; ir< mesh-2; ir += 2)
    {
        rhocg1 += 2 * aux(r[ir], rhoc[ir], gx, rab[ir]) + aux(r[ir+1], rhoc[ir+1], gx, rab[ir+1]);
    }//ir
    f_2 = aux(r[mesh-2], rhoc[mesh-2], gx, rab[mesh-2]);
    f_1 = aux(r[mesh-1], rhoc[mesh-1], gx, rab[mesh-1]);
    rhocg1 += f_2+f_2;
    rhocg1 += rhocg1;
    rhocg1 += f_0 + f_1;
    rhocg1/=3.0;

    drhocg [idx] = FOUR_PI / omega * rhocg1;
```

however, I'm not sure if the lambda expression will introduce additional overhead from function calls, which would need to be profiled to know for sure. If the kernel function itself isn't very time-consuming, your code is acceptable, although there is some efficiency loss.